File: src/entity/entity_bomb.c
#include "entity.h"

#include "../renderer/renderer.h"

static constexpr SDL_Rect frame[] = {
  {265, 265, 9, 21},
};

Entity *NewBomb() {
  auto e = NewEntity(EntityBomb);

  e->texture = 0;
  e->data = 1;

  e->pos.x = 0;
  e->pos.y = 0;
  e->pos.w = frame[0].w * 2;
  e->pos.h = frame[0].h * 2;

  e->xa = 0;
  e->ya = 0;

  e->damage = 1000;
  e->tickTime = 1;

  return e;
}

static void BombCalcSize(Entity *e, i32 *w, i32 *h) {
  const auto scale = pow(0.90, e->tickTime);

  *w = (i32)(e->pos.w * scale);
  *h = (i32)(e->pos.h * scale);

  if (*w == 0 || *h == 0) {
    e->data = 0;
    *w = 1;
    *h = 1;
  }
}

void BombTick(Entity *e) {
  e->tickTime += e->data;

  e->pos.x += e->xa * 2;
  e->pos.y += e->ya * 2;

  if (e->data != 0) {
    return;
  }

  SDL_Rect scaledrect;
  scaledrect.x = e->pos.x;
  scaledrect.y = e->pos.y;
  BombCalcSize(e, &scaledrect.w, &scaledrect.h);

  Entity *nextEntity = nullptr;
  for (auto it = entities; it != nullptr; it = nextEntity) {
    nextEntity = it->next;

    if (it->type == EntityShip && SDL_HasIntersection(&it->pos, &scaledrect)) {
      HurtEntity(it, e->damage);
      NewExplosion(scaledrect.x, scaledrect.y);

      if (player->score + 1 > player->score) {
        player->score++;
      }

      break;
    }
  }

  e->removed = true;
}

void BombRender(Entity *e) {
  const auto f = &frame[0];

  i32 width, height;
  BombCalcSize(e, &width, &height);

  RenderSprite(
    e->texture, e->pos.x, e->pos.y, width, height, f->x, f->y, f->w, f->h);
}

File: entity_bomb.go
package main

import (
	"math"

	"github.com/veandco/go-sdl2/sdl"
)

// bombFrames are the sprite frames for bombs.
var bombFrames = []sdl.Rect{
	{X: 265, Y: 265, W: 9, H: 21},
}

// NewBomb creates a new bomb entity at the player's position.
func NewBomb() *Entity {
	e := NewEntity(EntityTypeBomb)

	e.TicksDelta = 1

	frame := bombFrames[0]
	e.Pos = sdl.Rect{X: 0, Y: 0, W: frame.W * 2, H: frame.H * 2}

	e.Damage = 1000
	e.Ticks = 1

	return e
}

// bombCalcSize calculates the current size of a shrinking bomb.
// The bomb shrinks over time as it falls.
func bombCalcSize(e *Entity) (w, h int32) {
	scale := math.Pow(0.90, float64(e.Ticks))

	w = int32(float64(e.Pos.W) * scale)
	h = int32(float64(e.Pos.H) * scale)

	if w == 0 || h == 0 {
		e.TicksDelta = 0
		w = 1
		h = 1
	}

	return w, h
}

// BombTick moves the bomb and checks for collisions with ships/submarines.
func BombTick(e *Entity) {
	e.Ticks += e.TicksDelta

	e.Pos.X += e.VelX
	e.Pos.Y += e.VelY

	if e.TicksDelta != 0 {
		return
	}

	scaledRect := sdl.Rect{X: e.Pos.X, Y: e.Pos.Y}
	scaledRect.W, scaledRect.H = bombCalcSize(e)

	dead := false

	for i := range EntityPool {
		it := &EntityPool[i]

		canAttack := it.Kind == EntityTypeShip || it.Kind == EntityTypeSubmarine

		if !canAttack {
			continue
		}

		if !it.Pos.HasIntersection(&scaledRect) {
			continue
		}

		// Bomb hits the target.
		dead = true
		it.Hurt(e.Damage)

		NewExplosion(scaledRect.X, scaledRect.Y)

		switch it.Kind {
		case EntityTypeShip:
			player.Score += 50
		case EntityTypeSubmarine:
			player.Score += 200
		}
		break
	}

	if dead {
		e.Remove()
		return
	}
}

// BombRender draws the bomb with a shrinking animation.
func BombRender(e *Entity) {
	w, h := bombCalcSize(e)
	RenderSprite(e.Texture, sdl.Rect{X: e.Pos.X, Y: e.Pos.Y, W: w, H: h}, bombFrames[0])
}
